### PR TITLE
Correct setting of target and relatedTargets post-dispatch

### PR DESCRIPTION
**Pull request by annevk**
@whatwg/components this fixes the following:

* Adds "touch target list" as a concept hold by event, so touch events can be retargeted. (This will require some special handling for Touch objects, but I think it's workable.)
* It now resets target/relatedTarget/touch target list post-dispatch based on pre-dispatch conditions.
* It makes dispatch account for non-nodes.

(The one thing remaining here is having a good look at event constructors and making sure we can make that work for the relatedTargets design. I think that requires some kind of callback type thing.)

***
<a href="https://whatpr.org/dom/585.html" title="Last updated on Mar 27, 2018, 10:30 AM GMT (67e64b4)">Preview</a> | <a href="https://whatpr.org/dom/585/5be84cc...67e64b4.html" title="Last updated on Mar 27, 2018, 10:30 AM GMT (67e64b4)">Diff</a>

### CODE REVIEW

**Comment by hayatoito** [comment]
I would like to take a look, however, I will take a vacation for a few days. Let me review in the next week.

File: dom.bs [hunk 1]
```diff
@@ -1138,15 +1137,22 @@ for discussion).
...
+
+ <li><p>Let <var>relatedTargets</var> be a new <a for=/>list</a>.
+
+ <li><p><a for=list>For each</a> <var>relatedTarget</var> of <var>event</var>'s
+ <a for=Event>relatedTargets</a>, <a for=list>append</a> the result of <a>retargeting</a>
+ <var>relatedTarget</var> against <var>target</var> to <var>relatedTargets</var>.
 
- <li><p>If <var>target</var> is <var>relatedTarget</var> and <var>target</var> is not
- <var>event</var>'s <a for=Event>relatedTarget</a>, then return true.
+ <li><p>If <var>relatedTargets</var> <a for=list>contains</a> <var>target</var> and
```
Comment (smaug----) [inline dom.bs]: I doubt this is quite right in Touch case. Need to think this case some more.
I have some such case in mind where one of the relatedTargets does contain target before retargeting and then retargeting makes also some other relatedTarget to be target.

File: dom.bs [hunk 2]
```diff
@@ -1197,11 +1205,12 @@ for discussion).
...
 
      <li><p><a>Append to an event path</a> with <var>event</var>, <var>parent</var>, null,
-     <var>relatedTarget</var>, and <var>slot-in-closed-tree</var>.
+     <var>relatedTargets</var>, and <var>slot-in-closed-tree</var>.
     </ol>
 
-   <li><p>Otherwise, if <var>parent</var> and <var>relatedTarget</var> are identical, then set
-   <var>parent</var> to null.
+   <li><p>Otherwise, if <var>relatedTargets</var> <a for=list>contains</a> <var>parent</var>, then
+   set <var>parent</var> to null.
```
Comment (smaug----) [inline dom.bs]: Gut feeling is that also this has similar problem.

File: dom.bs [hunk 3]
```diff
@@ -5751,9 +5757,21 @@ is an object or one of its <a>shadow-including ancestors</a>.
 <var>B</var>, repeat these steps until they return an object:</p>
 
 <ol>
- <li><p>If <var>A</var>'s <a for=tree>root</a> is not a <a for=/>shadow root</a>, or <var>A</var>'s
- <a for=tree>root</a> is a <a>shadow-including inclusive ancestor</a> of <var>B</var>, then return
- <var>A</var>.
+ <li><p>If <var>A</var> is a {{Window}} object, then set <var>A</var> to <var>A</var>'s
+ <a>associated <code>Document</code></a>.
```
Comment (smaug----) [inline dom.bs]: Hmm, retargeting Window gives Document? Shouldn't this just return Window

Comment (annevk) [reply dom.bs]: Hmm yeah, I guess we just want it to behave as if you passed document.

Comment (annevk) [reply dom.bs]: I guess I should just remove this step altogether then?

File: dom.bs (hunk 1 above)
Comment (annevk) [reply dom.bs]: It seems https://github.com/w3c/web-platform-tests/blob/master/touch-events/touch-retargeting.html can be extended somehow for more test coverage for touch. It's not entirely clear to me how that works given that on desktop the events are not really exposed.

File: dom.bs (hunk 3 above)
Comment (TakayoshiKochi) [reply dom.bs]: Because window is not an instance of a Node, it falls into the line 5767 (A is not a node)?
Sounds good  - but wondering what's the reasoning behind adding this line.

File: dom.bs [hunk 4]
```diff
@@ -5751,9 +5757,21 @@ is an object or one of its <a>shadow-including ancestors</a>.
...
+
+ <li>
+  <p>If one of the following is true
+
+  <ul class=brief>
+   <li><var>A</var> is not a <a>node</a>
+   <li><var>B</var> is not a <a>node</a>
```
Comment (TakayoshiKochi) [inline dom.bs]: Basically the change for retargeting algorithm from the original is these 2 lines above, am I understanding correct?

File: dom.bs (hunk 3 above)
Comment (annevk) [reply dom.bs]: I added this because of https://github.com/whatwg/dom/issues/580#issuecomment-369813057, but I didn't really think it through.

File: dom.bs (hunk 4 above)
Comment (annevk) [reply dom.bs]: Yeah.

File: dom.bs (hunk 3 above)
Comment (hayatoito) [reply dom.bs]: I think removing this step is not enough. 
e.g. If A is a node in a shadow tree and B is a Window, we should loop until A reaches the outer most shadow host.

The following should work, I think.

1. If one of the following is true

   - A is not a node
   - A’s root is not a shadow root
   - B is a node and A’s root is a shadow-including inclusive ancestor of B

   then return A.

2. Set A to A’s root’s host.

Comment (annevk) [reply dom.bs]: Thanks, that does look good.

File: dom.bs (hunk 2 above)
Comment (hayatoito) [reply dom.bs]: Yup, I confirmed the Blink's implementation.
We need to distinguish *relatedTarget* and *relatedTargets* somehow because only *relatedTarget* (in the sense of before this PR) should be considered here.
"*relaterTagets* contains parent" causes a sort of *false positive*.

Comment (annevk) [reply dom.bs]: How should touch targets be adjusted though?

Comment (annevk) [reply dom.bs]: Since the feedback I got before was that whenever we have relatedTarget handling, we also need to consider touch target handling.

Comment (hayatoito) [reply dom.bs]: These steps always need some time to understand its intention exactly. :(

My intention is:
- All touch targets should be adjusted in the same way as relatedTarget should be adjusted. They both should be adjusted.
- However, there is one difference between relatedTarget and touch targets. Only relatedTarget should be used to decide to exit this while loop "11. While parent is non-null". This step sets *parent* to null, thus I think this is an attempt to exit the loop.
To decide to exit this loop or not, in other words, to decide whether an event should be stopped at a shadow root or not,  only relatedTarget should be used as its judgement. Touch targets shouldn't be used there.

Since TouchEvent doesn't look to have a relatedTarget to me, TouchEvent never set parent to null here. In other words, TouchEvent is never stopped at a shadow root even if its target and one of touch targets are same.

Comment (annevk) [reply dom.bs]: Okay, so we effectively need to have relatedTarget as an internal slot and touchTargets as an internal slot and most of the time we do the same thing for both, but not there and not at the other place @smaug---- pointed out (which is hidden now for some reason).

I can make that change. I'd love some help with tests for touch targets btw since I don't really have a good way to run them here.

Comment (annevk) [reply dom.bs]: (The alternative option is that we keep one slot but branch on event interface, but that seems a little icky, but let me know if you'd prefer that.)

Comment (hayatoito) [reply dom.bs]: I would love to help to write more tests, though I'm afraid I can write tests only for synthetic TouchEvents. I am not sure how to write a test for non-synthetic TouchEvents.

> (The alternative option is that we keep one slot but branch on event interface, but that seems a little icky, but let me know if you'd prefer that.)

I don't have any preference.

btw, I am wondering how these *relatedTargets* will be used in TouchEvent spec.
TouchEvents has 3 *relatedTargets* lists.

```
    readonly        attribute TouchList touches;
    readonly        attribute TouchList targetTouches;
    readonly        attribute TouchList changedTouches;
```

Comment (annevk) [reply dom.bs]: I was thinking that since those lists are all immutable, a flattened list backing them is fine, but that only works if retargeting is consistent regardless of which list.

Comment (hayatoito) [reply dom.bs]: I see. Your assumption is correct. Retargeting is consistent regardless of which list.

**Comment by annevk** [comment]
The PR now separates relatedTarget and touch target list and only has the special cases for relatedTarget. I still need to define an event creation callback of sorts so we can define TouchTarget / FocusEvent et al properly. I was also thinking of incorporating #402 here since we'll have to test that case anyway.

File: dom.bs [hunk 5]
```diff
@@ -1156,15 +1167,26 @@ for discussion).
...
+ <a for=Event>relatedTarget</a>, or an {{EventTarget}} object in <var>event</var>'s
+ <a for=Event>touch target list</a> is a <a>node</a> and its <a for=tree>root</a> is a
+ <a for=/>shadow root</a>, and false otherwise.
+
+ <li><p>Let <var>relatedTarget</var> be the result of <a>retargeting</a> <var>event</var>'s
+ <a for=Event>relatedTarget</a> against <var>target</var>.
 
  <li><p>If <var>target</var> is <var>relatedTarget</var> and <var>target</var> is not
  <var>event</var>'s <a for=Event>relatedTarget</a>, then return true.
+ <!-- XXX Since we return early, we might want to unset more flags here. See #402. -->
```
Comment (annevk) [inline dom.bs]: Should we also unset target et al here if clearTargetsPostDispatch is true? It seems that would be consistent... (If that's the case it seems we should turn this into a short algorithm that we invoke here and later on.)

Comment (hayatoito) [reply dom.bs]: I have never thought about this case seriously so far. In this early return, I assume that any event listener is never called.

In this case, we don't have to worry about any leaking of a node in a shadow tree via an event object. Users don't have any change to access an event object, other than the original event dispatcher.

btw, the consistency may matter. I'm fine to reset target, relatedTarget, and touch target list here too.

Comment (annevk) [reply dom.bs]: Yeah, I'm mostly wondering about it from a consistency perspective. I could see you'd have some custom dispatching code and later store the event object after dispatch in a global place and maybe you don't want things to be leaked there ever. Doesn't seem like a sound setup, but other than this case we do provide the guarantee that nothing would leak.

Comment (annevk) [reply dom.bs]: I've now made this change.

**Comment by annevk** [comment]
I consider this ready apart from the event construction specification callback, which is being discussed in #414. The ongoing feedback here is much appreciated though, please keep it coming.

File: dom.bs [hunk 6]
```diff
@@ -10070,6 +10112,7 @@ Tab Atkins,
 Takashi Sakamoto,
 Takayoshi Kochi,
 Theresa O'Connor,
+Theodore Dubois,
```
Comment (smaug----) [inline dom.bs]: Is this related to this pr?

Comment (annevk) [reply dom.bs]: Yes: https://github.com/whatwg/dom/issues/392.

File: dom.bs [hunk 7]
```diff
@@ -1156,149 +1166,160 @@ for discussion).
...
+     <li><p>Set <var>slot-in-closed-tree</var> to false.
+    </ol>
 
-  <ol>
-   <li><p>Set <var>event</var>'s {{Event/target}} attribute to the <b>target</b> of the last tuple
-   in <var>event</var>'s <a for=Event>path</a>, that is either <var>tuple</var> or preceding
-   <var>tuple</var>, whose <b>target</b> is non-null.
+   <li><p>Set <var>event</var>'s {{Event/eventPhase}} attribute to {{Event/CAPTURING_PHASE}}.
 
-   <li><p>Set <var>event</var>'s <a for=Event>relatedTarget</a> to <var>tuple</var>'s
```
Comment (smaug----) [inline dom.bs]: So where is this retargeting happening now?

File: dom.bs [hunk 8]
```diff
@@ -1338,6 +1360,18 @@ for discussion).
 <ol>
  <li><p>If <var>event</var>'s <a>stop propagation flag</a> is set, then return.
 
+ <li><p>Set <var>event</var>'s {{Event/target}} attribute to the <a for=Event/path>target</a> of the
+ last tuple in <var>event</var>'s <a for=Event>path</a>, that is either <var>tuple</var> or
```
Comment (smaug----) [inline dom.bs]: what does 'last' mean in this context? In which direction is that

File: dom.bs [hunk 9]
```diff
@@ -1338,6 +1360,18 @@ for discussion).
...
+ preceding <var>tuple</var>, whose <a for=Event/path>target</a> is non-null.
+
+ <li><p>Set <var>event</var>'s <a for=Event>relatedTarget</a> to <var>tuple</var>'s
+ <a for=Event/path>relatedTarget</a>.
+
+ <li><p>Set <var>event</var>'s <a for=Event>touch target list</a> to <var>tuple</var>'s
```
Comment (smaug----) [inline dom.bs]: hmm, perhaps a bit confusing, but I guess should be enough for touch objects. They just need to have a pointer to the event, so that they can update their target form there (basically map their original target to retargeted target or so)

**Comment by hayatoito** [comment]
Looks okay to me. Thank you for updating this!

**Comment by annevk** [comment]
So thinking about this in light of the tests at `shadow-dom/event-post-dispatch.html` it seems this is still not entirely correct. With the patch is proposed, you can observe whether an event started out in a shadow tree or not.

I think _clearTargets_ should instead be based on event's path. You find the last tuple in event's path with a non-null target. And then if that tuple contains a node inside a shadow tree in one of the three target fields, you set _clearTargets_ to true.

**Comment by annevk** [comment]
Hmm, and should https://whatpr.org/dom/585.html#concept-event-listener-invoke be updated to check the propagation flag later, so target et al still get updated? That would lead to problems too otherwise I think.

**Comment by hayatoito** [comment]
Does this need another review? Please ping me if the PR is ready.

**Comment by annevk** [comment]
@hayatoito I'd like input from you and @smaug---- on https://github.com/whatwg/dom/pull/585#issuecomment-374604473 and what a solution to that might look like. The other thing we need here are more tests I suppose.

**Comment by smaug----** [comment]
I think that approach sounds good. So, in case all the targets after actual dispatch are in light DOM, targets wouldn't get cleared.

**Comment by hayatoito** [comment]
> I think clearTargets should instead be based on event's path. You find the last tuple in event's path with a non-null target. And then if that tuple contains a node inside a shadow tree in one of the three target fields, you set clearTargets to true.

I see. This sounds a reasonable change. Nice catch.

**Comment by annevk** [comment]
Tentative commit message:
```
Shadow: fix event dispatch

This makes a number of large changes to event dispatch:

* The algorithm can no longer return early with true. The observable
  aspect is that the event object is "uninitialized" and false might 
  be returned if the event object was already canceled.
* It accounts for Touch Events in shadow trees by adding the 
  "touch target list" which is somewhat similar to relatedTarget.
* Retargeting is now solely controlled by the event dispatch 
  algorithm. The hook for other specifications is removed.
* Retargeting now properly accounts for non-node objects.
* Dispatch in general accounts for non-node objects too now.
* If targets need to be unset to prevent exposure of a shadow tree
  this is now calculated prior to invoking listeners.

Initializing relatedTarget and "touch target list" is up to the event
classes that need it. #614 will add "event constructing steps" that
enable this.

Tests: https://github.com/w3c/web-platform-tests/pull/9919
(probably still needs work given today's changes)

This closes #402, fixes #561, fixes #562, fixes #580, and fixes #602.
```
So more review is welcome here (for the last two commits and this commit message in particular) and more tests.

Sorry it got so big, but I'm not sure this could be elegantly made smaller.

**Comment by hayatoito** [comment]
I've reviewed it, and it looks okay to me. A really great work!

**Comment by annevk** [comment]
Bugs:

* https://bugs.webkit.org/show_bug.cgi?id=184079
* https://bugs.chromium.org/p/chromium/issues/detail?id=826702
* https://bugzilla.mozilla.org/show_bug.cgi?id=1449560

(No bug for Edge since they haven't started yet as far as I'm aware.)
